### DISCUSSION TITLE
Insert human-friendly link via completion

### DISCUSSION THREAD

**Discussion by WhyNotHugo** [zk-org/zk#439] [Ideas]
### Verify

- [X] I searched the existing discussions for similar ideas

### Share your idea or feature request

When using markdown style links, the LSP will insert links in the style of:

    [xyz1_my_page_title](xyz1_my_page_title)

It would make a lot more sense to use:

    [my_page_title](xyz1)

I.e.: put the page title in the visible part of the link, and the id as the link target.

## Solution 1

Allow customising how links are generated by the LSP with a new `link_template` option. This is the most flexible, but might be an overkill

## Solution 2

Simply change the format used to the one proposed. Less flexible but it's a nicer experience, IMHO.

**Comment by tjex** [comment]
By "insert links" do you mean via lsp completion or via `ZkInsertLink`?

In any case, a markdown link requires `[title](path)` no? `id` couldn't be substituted for `path`?

And the link format options in the config file don't suffice / lead you to any preexisting solution?
i.e, use a custom template for building the link.

```toml
# MARKDOWN SETTINGS
[format.markdown]

# Format used to generate links between notes.
# Either "wiki", "markdown" or a custom template. Default is "markdown".
link-format = "markdown"
```

**Reply by WhyNotHugo** [reply]
I mean LSP completions.

> In any case, a markdown link requires [title](path) no? id couldn't be substituted for path?

I don't quite understand what you mean here.

> And the link format options in the config file don't suffice / lead you to any preexisting solution?
> i.e, use a custom template for building the link.

I was looking at `docs/config-lsp.md` so I didn't see this before. This _should_ do what I'm looking for, but I can't quite get it to work.

Using:

```
[format.markdown]
hashtags = true
link-format = "[{{title}}]({{metadata.id}})"
```

Inserts:

```
[title of note here]()
```

I.e.: `metadata.id` is replaced with an empty string.

I also tried `link-format = "[{{title}}]({{id}})"`, with the same result.

**Reply by tjex** [reply]
So, regarding the confusion, given the markdown link schema / signature: `[title](path)`, where `title` is the hyperlinked text, and `path` is the relative file path to the linked document.

You can't just use the `id` as the `path` portion of a markdown link.  Because it is not a file path. It is just the id.

It _will_ work if you link to notes in the same folder, but if you try to link to a note in a _different folder_, the link will not resolve. i.e, the note will not be found. 

Why `{{id}}` is not being accepted at all in the `link-format` field, I'm not sure...

I also get this: `[[]]` when setting:

```toml
[format.markdown]
link-format = "[[{{id}}]]"
```

`id` generation is setup and working as the filenames are using the id.

Also, `id` is not a metadata attribute by default. In order to get `metadata.id` you need to actually insert the `id` into a notes metadata by way of a template:

e.g, default.md
```markdown
---
title: {{title}}
id: {{id}}
---

# {{title}}

{{content}}
```

At this point, if you try to link to this note now with `[{{title}}]({{metadata.id}})`, then it will work. But as described above, this will end up with buggy / faulty links.

**Reply by WhyNotHugo** [reply]
> You can't just use the id as the path portion of a markdown link. Because it is not a file path. It is just the id.

The LSP happily follows links with an id-only just fine.

I'm aware that trying to generate html pages for these markdown file would generate broken links tho, so I'm considering renaming them to simply `{{id}}.md`.

> It will work if you link to notes in the same folder, but if you try to link to a note in a different folder, the link will not resolve. i.e, the note will not be found.

I keep all my notes in one directory, if that's what you mean. But fair warning, I think I'll have to use full paths.

---

I think that `{{id}}` not working might be a bug.

**Reply by tjex** [reply]
Just for the sake of saving confusion for others: the LSP will not just follow links with only an id by default. This is only working for you because your filenames are the id.

Here is an example:

```toml
[format.markdown]

# Format used to generate links between notes.
# Either "wiki", "markdown" or a custom template. Default is "markdown".
link-format = "[{{title}}]({{metadata.id}})"
```

`./test file - etkc.md`:

```text
  6 ---
  5 id: etkc
  4 ---
  3
  2 # test file
  1
7   [other](gn68) <- resolves fine.
```

`./gn68.md`
```text
    6 ---
    5 id: gn68
    4 ---
    3 # other
    2
    1
E 7   [test file](etkc)     ■ not found
```

**Reply by WhyNotHugo** [reply]
From what I can tell, the LSP is pretty flexible. If my note is called `gn68_streaming.md`, a link with just `gn68` works fine.

**Reply by tjex** [reply]
huh, true. Just renamed the above to `etkc_test-file.md`, and it does resolve now. Nice find and good to have it clarified. Mickael has seemingly sprinkled lots of hidden gems through `zk`.

**Reply by mickael-menu** [reply]
Here's the logic to resolve a note: https://github.com/zk-org/zk/blob/7ae8a39de7ed9d46d67271c0b15f6f18dfb267e7/internal/adapter/lsp/server.go#L524-L530

Although the comment seems outdated, [it doesn't match note titles anymore](https://github.com/zk-org/zk/pull/218), only path prefixes and unique path portions.
